airflow/jobs/scheduler_job.py

```diff
@@ -1245,6 +1252,9 @@ def _process_executor_events(self, session: Session = None) -> int:
 
                 self.processor_agent.send_callback_to_execute(request)
 
+            if i % CALLBACK_SEND_BATCH_SIZE == 0:
```

Comment:
Hmmm, I don't _think_ this will help actually.

This block/line 1253 doesn't get called that often, so I suspect what is happening is that the processor_manager side of the pipe is full, so trying to send _even a single byte_ might block until the other end reads, but it can't read because it is also trying to write.

And trying to heartbeat every time before writing would be slow!.

🤔

Reply:
One option could be for `send_callback_to_execute` to store in a queue internally in the ProcessorAgent, and only send inside heartbeat after first reading?

Reply:
(SLA callback requests also have this same problem.)

Reply:
It may be easier to change the other side - https://github.com/apache/airflow/blob/master/airflow/utils/dag_processing.py#L698 that is the _only_ place that side sends anything.

Reply:
Another option:

```python console
>>> p1, p2 = multiprocessing.Pipe()
>>> fcntl.fcntl(p1.fileno(), fcntl.F_SETFL, os.O_NONBLOCK | fcntl.fcntl(p1.fileno(), fcntl.F_GETFL)) 
>>> p1.send('a' * 1024000) # Simulate something that would block
BlockingIOError: [Errno 11] Resource temporarily unavailable
```

We could set the socket in DagFileManager (the receiving side) to non-blocking, that way if the send would fail we could catch it, and queue it there, and then go and poll again.?

Reply:
What do you mean by “queue it there, and then go and poll again”? Making the child connection non-blocking sounds great (I wondered briefly but couldn’t find found), but the polling loop seems to have quite a few side effects before that line, so we can’t just put the thing back and try again later. Does it mean to add a queue (list) on `DagFileProcessorManager` to store `DagParsingStat` instances when they fail to send, and add a block in the beginning of the poll loop to check and resend them first?